src/lib/parameters/flashparams/flashfs32.c

```diff
@@ -0,0 +1,1201 @@
+ * Name: parameter_flashfs_init
+ *
+ * Description:
+ *   This helper function advances the flash entry header pointer to the
+ *   locations of the next entry.
+ *
+ * Input Parameters:
+ *   fconfig      - A pointer to an null entry terminated array of
+ *                  flash_file_sector_t
+ *    buffer      - A pointer to a memory to make available to callers
+ *                  for write operations. When allocated to the caller
+ *                  space is reserved in the front for the
+ *                  flash_entry_header_t.
+ *                  If this is passes as NULL. The buffer will be
+ *                  allocated from the heap on calls to
+ *                  parameter_flashfs_alloc and fread on calls
+ *                  to parameter_flashfs_free
+ *
+ *   size         - The size of the buffer in bytes. Should be be 0 if buffer
+ *                  is NULL
+ *
+ * Returned value:
+ *                - A pointer to the next file header location
+ *
+ *
+ ****************************************************************************/
+
+int parameter_flashfs_init(sector_descriptor_t *fconfig, uint8_t *buffer, uint16_t size)
+{
+	int rv = 0;
+	sector_map = fconfig;
+	working_buffer_static = buffer != NULL;
+
+	if (!working_buffer_static) {
+		size = 0;
+	}
+
+	working_buffer = buffer;
+	working_buffer_size = size;
+
+	/* Sanity check */
+
+	flash_entry_header_t *pf = find_entry(parameters_token);
+
+	/*  No paramaters */
+
+	if (pf == NULL) {
+		size_t total_size = size + sizeof(flash_entry_header_t);
+		size_t alignment = 31;//32-byte flash line - 1
+		size_t  size_adjust = ((total_size + alignment) & ~alignment) - total_size;
+		total_size += size_adjust;
+
+		/* Do we have free space ?*/
+
+		if (find_free(total_size) == NULL) {
+
+			/* No parameters and no free space => need erase */
+
+			rv  = parameter_flashfs_erase();
+		}
+	}
+
+	return rv;
+}
+
+#if defined(FLASH_UNIT_TEST)
+
+static sector_descriptor_t  test_sector_map[] = {
+	{14, 128 * 1024, 0x081c0000},
+	{15, 128 * 1024, 0x081e0000},
+	{0, 0, 0},
+};
+
+__EXPORT void test(void);
+
+uint8_t test_buf[32 * 1024];
+
+__EXPORT void test(void)
+{
+	uint16_t largest_block = (32 * 1024) + 96;
+	uint8_t *buffer = malloc(largest_block);
+
+	parameter_flashfs_init(test_sector_map, buffer, largest_block);
+
+	for (int t = 0; t < sizeof(test_buf); t++) {
+		test_buf[t] = (uint8_t) t;
+	}
+
+	int er = parameter_flashfs_erase();
+	uint8_t *fbuffer;
+	size_t buf_size;
+	int written = 0;
+	int read = 0;
+	int rv = 0;
+
+	for (int a = 0; a <= 4; a++) {
+		rv =  parameter_flashfs_alloc(parameters_token, &fbuffer, &buf_size);
+		memcpy(fbuffer, test_buf, a);
+		buf_size = a;
+		written = parameter_flashfs_write(parameters_token, fbuffer, buf_size);
+		read = parameter_flashfs_read(parameters_token, &fbuffer, &buf_size);
+		parameter_flashfs_free();
+
+		if (read != written) {
+			static volatile int j;
+			j++;
+		}
+	}
+
+	int block = 2048;
```

Comment:
@taylornelms15 - did you run this with 2 sectors and verify the transition to the next sector?